server/bin/pbench-dispatch.sh

```diff
@@ -3,222 +3,284 @@
 sort ${list}.unsorted > ${list}
-rm -f ${list}.unsorted
+sts=${?}
+if [[ ${sts} != 0 ]]; then
+    log_exit "Failed: \"sort ${list}.unsorted > ${list}\", status ${sts}" 6
+fi
 
-typeset -i ntb=0
 typeset -i ntotal=0
+typeset -i ntbs=0
+typeset -i npartialsucc=0
 typeset -i nerrs=0
 typeset -i ndups=0
-typeset -i npartialsucc=0
+typeset -i nquarantined=0
 
-while read tarball ;do
-    let ntotal+=1
+while read tbmd5; do
+    ntotal=${ntotal}+1
 
-    linksrc_path=$(dirname $tarball)
-    tb_linksrc=$(basename $linksrc_path)
-    if [ "$linksrc" != "$tb_linksrc" ]; then
-        # All is NOT well: we expect $linksrc as the parent directory name
-        # of the symlink tar ball name.
-        log_exit "$TS: FATAL - unexpected \$linksrc for $tarball" 57
-    fi
+    # Full pathname of tar ball
+    tb=${tbmd5%.md5}
 
-    controller_path=$(dirname $linksrc_path)
-    controller=$(basename $controller_path)
-    if [ "$ARCHIVE" != "$(dirname $controller_path)" ]; then
-        # The controller's parent is not $ARCHIVE!
-        log_exit "$TS: FATAL - unexpected archive directory for $tarball" 57
-    fi
+    # Controller directory path
+    tbdir=$(dirname ${tb})
+
+    # resultname: get the basename foo.tar.xz and then strip the .tar.xz
+    resultname=$(basename ${tb})
+    resultname=${resultname%.tar.xz}
 
-    link=$(readlink -e $tarball)
-    if [ -z "$link" ] ;then
-        log_error "$TS: symlink target for $tarball does not exist" "${mail_content}"
-        let nerrs+=1
-        quarantine ${controller_path}/_QUARANTINED/BAD-LINK ${tarball}
+    # The controller hostname is the last component of the directory part of
+    # the full path
+    controller=$(basename ${tbdir})
+
+    dest=${ARCHIVE}/${controller}
+
+    if [[ -f ${dest}/${resultname}.tar.xz || -f ${dest}/${resultname}.tar.xz.md5 ]]; then
+        log_error "${TS}: Duplicate: ${tb} duplicate name" "${status}"
+        quarantine ${duplicates}/${controller} ${tb} ${tbmd5}
+        (( ndups++ ))
         continue
     fi
 
-    resultname=$(basename $tarball)
-    resultname=${resultname%.tar.xz}
+    pushd ${tbdir} > /dev/null 2>&4
+    md5sum --check ${resultname}.tar.xz.md5
+    sts=${?}
+    popd > /dev/null 2>&4
+    if [[ ${sts} -ne 0 ]]; then
+        log_error "${TS}: Quarantined: ${tb} failed MD5 check" "${status}"
+        quarantine ${quarantine}/${controller} ${tb} ${tb}.md5
+        (( nquarantined++ ))
+        continue
+    fi
 
-    # XXXX - for now, if it's a duplicate name, just punt and avoid
-    # producing the error
-    if [ ${resultname%%.*} == "DUPLICATE__NAME" ] ;then
-        let ndups+=1
-        quarantine ${controller_path}/_QUARANTINED/DUPLICATES ${link} ${link}.md5
-        rm -f ${tarball}
-        status=$?
-        if [ $status -ne 0 ] ;then
-            log_error "$TS: Cannot remove $tarball link: code $status" "${mail_content}"
-        fi
+    # Make sure that all the relevant state directories exist
+    mk_dirs ${controller}
+    sts=${?}
```

Comment:
Are we creating state directories here or controller directory?

Reply:
What @dbutenhof said ...